### PR TITLE
[Feature] Support multiple path

### LINKED ISSUE #355: Specify multiple files/folders on input
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| Library version  | master

Right now we support supplying a folder or directory, to specify where the tool should run on. However it does not support specifying multiple files/folders.

I propose a syntax like so
```
php insights analyse first/folder first/randomfile.php /second/folder
```

Where I see the most problems with this change is all the places we use `Configuration::getDirectory`, for example in the formatters to shorten the names. Our `FilesRepository:within` would also have to change, to now support an array of paths.

### PR DESCRIPTION
**Pull request by 50bhan**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Fixed tickets | #355 

This PR will close #355 for support of multiple files/folders on input.

There is couple of thing that should be considered:
- First, some parts of this PR is based on @Jibbarth awesome PR #382. So, I assumed his PR will be merged first.
- Secondly, there are some discussable points in this code, so we should discuss about them.

### CODE REVIEW

File: src/Application/ConfigResolver.php [hunk 1]
```diff
@@ -46,8 +46,8 @@ final class ConfigResolver
     public static function resolve(array $config, InputInterface $input): Configuration
     {
         $directory = DirectoryResolver::resolve($input);
-        $config = ConfigResolver::mergeInputRequirements($config, $input);
-        $composer = self::getComposer($directory);
+        $config = self::mergeInputRequirements($config, $input);
+        $composer = self::getComposer($directory[0]);
```
Comment (olivernybroe) [inline src/Application/ConfigResolver.php]: hmm, so I see you choose that the composer file must be from the first directory which has been added.

I kinda get that, I don't think we will benefit much from allowed multiple composer files. 

The only thing I am worried about is how intuitive this will be for the user.

Comment (olivernybroe) [reply src/Application/ConfigResolver.php]: Would could always allow a new parameter, to specify the composer path and if that is not set, then we just fallback to looking at the first file supplied.

File: src/Application/Console/Formatters/Checkstyle.php [hunk 2]
```diff
@@ -57,7 +57,7 @@ public function format(
 
                 /** @var Details $detail */
                 foreach ($details as $detail) {
-                    $fileName = $this->getFileName($detail, $dir);
+                    $fileName = $this->getFileName($detail, $dir[0]);
```
Comment (olivernybroe) [inline src/Application/Console/Formatters/Checkstyle.php]: Hmm this ony is a bit more tricky. Don't think we can use the first dir here for it. 
As then it would shorten the filename wrongly. 

So we need to think of an approach for shortening the filename when you supply multiple paths.
This is a problem for almost all formatters

File: src/Domain/Insights/InsightCollectionFactory.php [hunk 3]
```diff
@@ -59,17 +59,17 @@ public function get(
...
             $files = array_map(static function (\SplFileInfo $file) {
                 return $file->getRealPath();
-            }, iterator_to_array($this->filesRepository->within($dir, $this->config->getExcludes())->getFiles()));
+            }, $this->filesRepository->within($directoryList, $this->config->getExcludes())->getFiles());
         } catch (\InvalidArgumentException $exception) {
             throw new DirectoryNotFound($exception->getMessage(), 0, $exception);
         }
 
-        $collector = $this->analyser->analyse($dir, $files);
+        $collector = $this->analyser->analyse($directoryList[0], $files);
```
Comment (olivernybroe) [inline src/Domain/Insights/InsightCollectionFactory.php]: Here we have again a scenario where we have to figure out the approach.
The dir supplied here is passed down to the `Collector`, in the collector it is used to shorten down the name of the file to only filename and not the full path.

File: src/Application/Console/Formatters/Checkstyle.php (hunk 2 above)
Comment (50bhan) [reply src/Application/Console/Formatters/Checkstyle.php]: @olivernybroe Yes you are right! This is one of the points I wanted to discuss. Actually there is three points like this which i'm concern about.

File: src/Application/Console/Formatters/Console.php [hunk 4]
```diff
@@ -329,7 +331,7 @@ private function issues(
 
                 /** @var \NunoMaduro\PhpInsights\Domain\Details $detail */
                 foreach ($details as $detail) {
-                    $detailString = $this->formatFileLine($detail, $dir, $category);
+                    $detailString = $this->formatFileLine($detail, '', $category);
```
Comment (50bhan) [inline src/Application/Console/Formatters/Console.php]: @olivernybroe This line is the same as those two scenarios.

File: src/Application/Console/Formatters/Checkstyle.php (hunk 2 above)
Comment (olivernybroe) [reply src/Application/Console/Formatters/Checkstyle.php]: Alright, so shortening the filename. 

So let's paint a scenario, I give php insights 4 paths, because I modified these 4 files
- `app/Console/Commands/MyTestCommand.php`
- `/Users/olivernybroe/Code/myCodeBase/app/Http/Controllers/Auth/LoginController.php`
- `routes/api.php`
- `tests/Feature/Auth/LoginControllerTest.php`

When I see the output from these files, shortening them based on the first input would result in all except the first one showing the full path (e.g `/Users/olivernybroe/Code/myCodeBase/app/Http/Controllers/Auth/LoginController.php`) which is not ideal.  
I think the most ideal shortening that we could do for this scenario would be to remove `/Users/olivernybroe/Code/myCodeBase`from all files. So basically finding the most folders which they all have in common.

Let's look at another example with files in different projects
- `/Users/olivernybroe/Code/myCodeBase/app/Http/Controllers/Auth/LoginController.php`
- `/Users/olivernybroe/Code/myOtherCodeBase/bin/phpinsights.php`

In this case the best shortening would be `/Users/olivernybroe/Code`.

So basically what we need to do is:
- Take all the paths the user has given us and get the full path for them all
- Sort the paths
- Take the first and last element of the paths
- Compare first and last element, to find longest common path

This works because we sort the paths and if we only have one path, then first and last element is just the same. When comparing the first and last element we prob. have to take into account the path separator. Because if we look at my second example, the first characters that are in common is actually `/Users/olivernybroe/Code/my`, however it should only remove `/Users/olivernybroe/Code` from the paths.


@50bhan What do you think of this approach?

**Comment by nunomaduro** [comment]
I think we should target the 2.x version.

**Comment by 50bhan** [comment]
@nunomaduro If it's the case, i'm cool with that! Maybe we'll be able to add some related features to it as well (i.e. #354)

File: src/Application/ConfigResolver.php (hunk 1 above)
Comment (50bhan) [reply src/Application/ConfigResolver.php]: @olivernybroe Yeah I thought it's simpler to get composer file from first directory (specially when we run analyser without specifying any path). And adding a parameter is nice and I can go for it.

File: src/Application/Console/Formatters/Checkstyle.php (hunk 2 above)
Comment (50bhan) [reply src/Application/Console/Formatters/Checkstyle.php]: @olivernybroe It's a nice approach and i'm agree with it. Specially I don't like shortening to file name because what if we had two files with same name at different paths? So, I think this works. In case of one path, we can shortening to filename because there will be no conflict in names. In other cases we will shortening by common path which should be calculated by path separator as you mentioned.

Comment (olivernybroe) [reply src/Application/Console/Formatters/Checkstyle.php]: Sounds great. Let's go with this approach then!

Exactly, with one path, we shorten to filename, and this approach should actually result in that, as the longest common path when we only have one path, is basically the full path.

**Comment by olivernybroe** [comment]
@50bhan Adding support for this, will make it a lot easier to add support for #354. However that should be in a pr for itself :)

**Comment by 50bhan** [comment]
@olivernybroe Yeah sure it is. Honestly I hate waiting too much for tests to pass or FAIL :laughing: So, I took this task so #354 will be possible after it. And of course in a separate PR.

**Comment by 50bhan** [comment]
@olivernybroe Sorry to bother, I did those changes and I will be appreciate if you take a look at it. Thanks.

File: src/Application/Console/Formatters/PathShortener.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ * @internal
+ */
+final class PathShortener
+{
+    /**
+     * @param array<string> $paths
+     *
+     * @return string
+     */
+    public static function extractCommonPath(array $paths): string
```
Comment (olivernybroe) [inline src/Application/Console/Formatters/PathShortener.php]: My solution was more something like this. Don't know if you find my solution simpler or not :)

```php
        $paths = array_values($paths);
        sort($paths);

        $first = $paths[0];
        $last = $paths[count($paths)-1];
        $length = min(strlen($first), strlen($last));

        for ($index=0; $index<$length && $first[$index]==$last[$index]; $index++);

        $prefix = mb_substr($first, 0, $index ?? 0);

        return mb_substr(
            $prefix,
            0,
            mb_strrpos($prefix, DIRECTORY_SEPARATOR)+1
        );
```

Comment (50bhan) [reply src/Application/Console/Formatters/PathShortener.php]: @olivernybroe Yes your solution is simpler and I like it more. Thanks a lot :)

**Comment by 50bhan** [comment]
@olivernybroe Thank you for fixing the bug and test. I imagine there is nothing more for this task. Let me know if something came up regarding this PR.

**Comment by olivernybroe** [comment]
@50bhan Just merged jibbarths speed improvements, can you resolve the conflicts quickly? 👍

**Comment by 50bhan** [comment]
@olivernybroe Yeah i'm on it.

**Comment by olivernybroe** [comment]
Hey @Jibbarth can I get your to make a review of this? 

I have tested it locally and it works 😄

**Comment by Jibbarth** [comment]
@olivernybroe Sure, I'll check this tonight :+1:

File: docs/get-started.md [hunk 6]
```diff
@@ -77,6 +77,14 @@ In laravel, launch command as usual with your path:
 php artisan insights path/to/analyse
 ```
 
+## Specify composer file
```
Comment (Jibbarth) [inline docs/get-started.md]: Should we add a badge about v2 version on this ?

File: src/Application/ConfigResolver.php [hunk 7]
```diff
@@ -62,7 +64,7 @@ public static function resolve(array $config, InputInterface $input): Configurat
 
         $isRootAnalyse = true;
         foreach (Kernel::getRequiredFiles() as $file) {
-            if (! file_exists($directory . DIRECTORY_SEPARATOR . $file)) {
+            if (! file_exists($directories[0] . DIRECTORY_SEPARATOR . $file)) {
                 $isRootAnalyse = false;
```
Comment (Jibbarth) [inline src/Application/ConfigResolver.php]: Well, if I remember correctly, `$isRootAnalyse` is used to exclude some global insights related to composer. Kernel::getRequiredFiles return [composer.json, composer.lock] 

As you manage the possibility to pass a custom composer.json, maybe it should be handle here ?

File: src/Application/ConfigResolver.php [hunk 8]
```diff
@@ -154,11 +156,23 @@ private static function mergeInputRequirements(array $config, InputInterface $in
...
+     * @return Composer
+     */
+    private static function getComposer(InputInterface $input, array $directories): Composer
     {
-        $composerPath = $directory . DIRECTORY_SEPARATOR . 'composer.json';
+        /** @var string|null $composerPath */
+        $composerPath = $input->hasOption('composer') ? $input->getOption('composer') : null;
+
+        if ($composerPath === null) {
+            $directory = rtrim($directories[0], '/');
```
Comment (Jibbarth) [inline src/Application/ConfigResolver.php]: As we use only $directories[0], maybe we should pass it directly to the function ?

File: src/Domain/Collector.php [hunk 9]
```diff
@@ -215,17 +219,17 @@ final class Collector
...
      *
-     * @param  string  $dir
+     * @param string $dir
+     * @param array<string> $files
      */
-    public function __construct(string $dir)
+    public function __construct(string $dir, array $files)
     {
         $this->dir = $dir;
+        $this->commonPath = $files !== [] ? PathShortener::extractCommonPath($files) : '';
```
Comment (Jibbarth) [inline src/Domain/Collector.php]: maybe we should pass the 'commonPath' instead of the array of files ?

File: src/Domain/ComposerLoader.php [hunk 10]
```diff
@@ -24,7 +24,7 @@ final class ComposerLoader
 
     public static function getInstance(Collector $collector): Composer
     {
-        if (null === self::$composer || $collector !== self::$currentCollector) {
+        if (self::$composer === null || $collector !== self::$currentCollector) {
```
Comment (Jibbarth) [inline src/Domain/ComposerLoader.php]: Why this change :sweat_smile: You don't like Yoda ? :wink:

File: src/Application/ConfigResolver.php [hunk 11]
```diff
@@ -73,7 +75,7 @@ public static function resolve(array $config, InputInterface $input): Configurat
         }
 
         if (! isset($config['directory'])) {
-            $config['directory'] = $directory;
+            $config['directory'] = $directories;
```
Comment (Jibbarth) [inline src/Application/ConfigResolver.php]: Maybe we should rename $config['directory'] into $config['directories'] as we now pass an array of directories ? 

And perhaps add an 
```
if (isset($config['directory']) {
   $config['directories'] = [$config['directory']];
}
```
for less breaking change ? WDYT ?

File: src/Domain/Insights/InsightCollection.php [hunk 12]
```diff
@@ -82,7 +82,7 @@ public function allFrom(Metric $metric): array
      */
     public function results(): Results
     {
-        if (null !== $this->results) {
+        if ($this->results !== null) {
```
Comment (Jibbarth) [inline src/Domain/Insights/InsightCollection.php]: Again :sweat_smile:

File: src/Domain/Insights/InsightCollectionFactory.php [hunk 13]
```diff
@@ -59,17 +59,17 @@ public function get(
...
             $files = array_map(static function (\SplFileInfo $file) {
                 return $file->getRealPath();
-            }, $this->filesRepository->within($dir, $this->config->getExcludes())->getFiles());
+            }, $this->filesRepository->within($directories, $this->config->getExcludes())->getFiles());
         } catch (\InvalidArgumentException $exception) {
             throw new DirectoryNotFound($exception->getMessage(), 0, $exception);
         }
 
-        $collector = $this->analyser->analyse($dir, $files);
+        $collector = $this->analyser->analyse($directories[0], $files);
```
Comment (Jibbarth) [inline src/Domain/Insights/InsightCollectionFactory.php]: No need to analyse for each directory ? 

What about launch inspection on `src` and `tests` folders ? The analyse will be it just on `src` ?

File: src/Domain/Insights/InsightFactory.php [hunk 14]
```diff
@@ -118,7 +118,7 @@ private function runInsightCollector(OutputInterface $consoleOutput): void
     private function loadInsights(array $insights): array
     {
         $insightsAdded = [];
-        $cwd = getcwd() ?? $this->config->getDirectory();
+        $cwd = getcwd() ?? $this->config->getDirectories();
```
Comment (Jibbarth) [inline src/Domain/Insights/InsightFactory.php]: $cwd must be a string. It'll work in case getcwd return false ?

File: src/Infrastructure/Repositories/LocalFilesRepository.php [hunk 15]
```diff
@@ -40,33 +44,76 @@ public function getDefaultDirectory(): string
...
+            ->in($this->fileList['dirname'])
+            ->name($this->fileList['basename']);
+
+        return $this->getIterator();
+    }
+
+    /**
+     * @return array<\Symfony\Component\Finder\SplFileInfo>
+     */
+    private function getIterator(): array
```
Comment (Jibbarth) [inline src/Infrastructure/Repositories/LocalFilesRepository.php]: I would named this getFilesList instead getIterator ?

File: src/Application/ConfigResolver.php (hunk 11 above)
Comment (olivernybroe) [reply src/Application/ConfigResolver.php]: We can allow breaking changes, as this is for V2 😄 
So maybe just make the breaking change and don't add the extra logic.

**Comment by 50bhan** [comment]
@Jibbarth Welcome :wink: Thanks for the review! Yeah I totally forgot about docs. I will add it and go for your comments.

File: src/Domain/ComposerLoader.php (hunk 10 above)
Comment (50bhan) [reply src/Domain/ComposerLoader.php]: hahaaa I have a thing for Yoda style :laughing: Just kidding. I was just trying to get more scores in architecture for insight to pass the tests! So I changed them.

File: src/Domain/Insights/InsightCollectionFactory.php (hunk 13 above)
Comment (50bhan) [reply src/Domain/Insights/InsightCollectionFactory.php]: We don't need directory path for analyser, because it just need to know what files should be analyse. We should pass some files path to analyser and it do the job. Directory path was mainly for shortening path in `Collector`. So basically I can even remove the $directories[0] from analyse function. But this first index in a particular situation is useful (when user only want to analyse one file/dir). So I let it be.

So, if we launch inspection on `src` and `tests` folders, list of both directory files will be send to analyser and everything will be OK.

Comment (50bhan) [reply src/Domain/Insights/InsightCollectionFactory.php]: If you have any idea for further improvement, i'm up for that :+1:

File: src/Domain/Collector.php (hunk 9 above)
Comment (50bhan) [reply src/Domain/Collector.php]: We pass and array of files to `PathShortener` and it will extract common path and get back to us. In case of no files (just for safety and not breaking the tests) common path will be empty string.

File: docs/get-started.md (hunk 6 above)
Comment (50bhan) [reply docs/get-started.md]: Done

File: src/Infrastructure/Repositories/LocalFilesRepository.php (hunk 15 above)
Comment (50bhan) [reply src/Infrastructure/Repositories/LocalFilesRepository.php]: Done

File: src/Application/ConfigResolver.php (hunk 11 above)
Comment (50bhan) [reply src/Application/ConfigResolver.php]: Yeah I just did with without considering BC. As Oliver said it's for v2 and I think BC is fine. So if there is another matter just let me know.

File: src/Application/ConfigResolver.php (hunk 8 above)
Comment (50bhan) [reply src/Application/ConfigResolver.php]: Done

File: src/Application/ConfigResolver.php (hunk 7 above)
Comment (50bhan) [reply src/Application/ConfigResolver.php]: Nice catch! Done.

**Comment by olivernybroe** [comment]
@50bhan I found a bug when adding in a file which also exist in the vendor folder. 
So for example running `bin/phpinsights analyse CHANGELOG.md` results in 260 files found, because it finds a bunch in the vendor folder.

**Comment by 50bhan** [comment]
@olivernybroe I'm tuning the code a little bit, after that I will fix it as well.

**Comment by olivernybroe** [comment]
@50bhan Cool! 
Just a quick example of something we can do (might need some fine-tuning) to only run on changed files when this is merged in
```bash
bin/phpinsights analyse $(git --no-pager diff master HEAD --name-only -- '*.php' | tr '\n' ' ')
```

**Comment by 50bhan** [comment]
@olivernybroe @Jibbarth I did some new changes and tried to increase consistency and simplicity. Also I considered all of Jibé review points. So, do me a favor and check it out.

File: src/Domain/Insights/ForbiddenSecurityIssues.php [hunk 16]
```diff
@@ -70,7 +70,8 @@ private function getResult(): Result
 
             try {
                 self::$result = $checker->check(sprintf(
-                    '%s/composer.lock', $this->collector->getDir()
+                    '%scomposer.lock',
```
Comment (olivernybroe) [inline src/Domain/Insights/ForbiddenSecurityIssues.php:73]: Does commonPath include the backslash or did you forget to add it in here?

Comment (50bhan) [reply src/Domain/Insights/ForbiddenSecurityIssues.php:73]: Yes it contains it always.

File: docs/get-started.md [hunk 17]
```diff
@@ -77,7 +76,27 @@ In laravel, launch command as usual with your path:
 php artisan insights path/to/analyse
 ```
 
-## Specify composer file
+## Analyse multiple paths ![Version](https://img.shields.io/badge/Version-2.0.0-blue)
```
Comment (Jibbarth) [inline docs/get-started.md:79]: In others docs page, we use `<Badge text="^1.x"/>` for the version. I'm not against use a badge from shields.io, but we should stay consistency. 
WDYT @olivernybroe, should we use a badge from shields.io everywhere ?

File: src/Application/Console/Definitions/AnalyseDefinition.php [hunk 18]
```diff
@@ -17,9 +17,9 @@ public static function get(): InputDefinition
     {
         return new InputDefinition([
             new InputArgument(
-                'directory',
+                'directories',
                 InputArgument::OPTIONAL | InputArgument::IS_ARRAY,
-                'The directory to analyse'
+                'The directories to analyse'
```
Comment (Jibbarth) [inline src/Application/Console/Definitions/AnalyseDefinition.php:22]:
```suggestion
                'The directories or files to analyse'
```

As we can pass also file path. Maybe we should rename argument into `paths` instead `directories` ?

Comment (Jibbarth) [reply src/Application/Console/Definitions/AnalyseDefinition.php:22]: Could be done in another PR BTW

File: src/Domain/Collector.php [hunk 19]
```diff
@@ -219,13 +219,12 @@ final class Collector
...
-     * @param array<string> $files
+     * @param array<string> $directories
      */
-    public function __construct(string $dir, array $files)
+    public function __construct(array $directories)
     {
-        $this->dir = $dir;
-        $this->commonPath = $files !== [] ? PathShortener::extractCommonPath($files) : '';
+        $this->analysedDirectories = $directories;
+        $this->commonPath = PathShortener::extractCommonPath($directories);
```
Comment (Jibbarth) [inline src/Domain/Collector.php:227]: The `extractCommonPath` has be already done in `ConfigResolver`. We should use Configuration to retrieve it to avoid twice call ?

File: src/Domain/Insights/InsightFactory.php [hunk 20]
```diff
@@ -118,15 +118,14 @@ private function runInsightCollector(OutputInterface $consoleOutput): void
...
-        $cwd = getcwd() ?? $this->config->getDirectories();
 
         foreach ($insights as $insight) {
             /** @var InsightLoader $loader */
             foreach ($this->insightLoaders as $loader) {
                 if ($loader->support($insight)) {
                     $insightsAdded[] = $loader->load(
                         $insight,
-                        (string) $cwd,
+                        (string) (getcwd() ?? $this->config->getCommonPath()),
```
Comment (Jibbarth) [inline src/Domain/Insights/InsightFactory.php:128]: Can we keep this outside the double foreach ? It should not change on each loop.

File: src/Application/Console/Formatters/GithubAction.php [hunk 21]
```diff
@@ -39,19 +39,22 @@ public function __construct(InputInterface $input, OutputInterface $output)
     {
         $this->decorated = new Console($input, $output);
         $this->output = $output;
-        $this->baseDir = Container::make()->get(Configuration::class)->getDirectory();
+        $this->baseDir = Container::make()->get(Configuration::class)->getDirectories();
```
Comment (Jibbarth) [inline src/Application/Console/Formatters/GithubAction.php:42]: What about use commonPath ?

File: docs/get-started.md (hunk 17 above)
Comment (olivernybroe) [reply docs/get-started.md:79]: I think we should stay with `Badge` at least for this pr. 

Using shield.io badges I think would be too heavy, as it would have to load a bunch of images, when a simple html element can be used instead.

Comment (50bhan) [reply docs/get-started.md:79]: Ok, so I will change it to HTML badge.

File: src/Domain/Insights/InsightFactory.php (hunk 20 above)
Comment (50bhan) [reply src/Domain/Insights/InsightFactory.php:128]: Yes, Thanks! Done.

File: src/Application/Console/Formatters/GithubAction.php (hunk 21 above)
Comment (50bhan) [reply src/Application/Console/Formatters/GithubAction.php:42]: Agree :+1: Done.

File: src/Domain/Collector.php (hunk 19 above)
Comment (50bhan) [reply src/Domain/Collector.php]: Done

File: src/Application/Console/Definitions/AnalyseDefinition.php (hunk 18 above)
Comment (50bhan) [reply src/Application/Console/Definitions/AnalyseDefinition.php:22]: Yep I agree. I try to do that in current PR.

**Comment by 50bhan** [comment]
@olivernybroe @Jibbarth I think we are good to go. Final check? :laughing:

File: src/Application/Console/Definitions/AnalyseDefinition.php [hunk 22]
```diff
@@ -17,9 +17,9 @@ public static function get(): InputDefinition
     {
         return new InputDefinition([
             new InputArgument(
-                'directories',
+                'paths',
                 InputArgument::OPTIONAL | InputArgument::IS_ARRAY,
-                'The directories to analyse'
+                'The paths or files to analyse'
```
Comment (Jibbarth) [inline src/Application/Console/Definitions/AnalyseDefinition.php:22]:
```suggestion
                'Paths of directories or files to analyse'
```

**Comment by 50bhan** [comment]
@Jibbarth Thank you :) Actually this one you mentioned is fixed already https://github.com/nunomaduro/phpinsights/blob/25d73931b7afaf23248d441f9c818e8dc655a800/src/Application/Console/Definitions/AnalyseDefinition.php#L22

**Comment by Jibbarth** [comment]
@50bhan no it's not :sweat_smile: I suggested to replace `The paths or files to analyse` by `Paths of directories or files to analyse` :wink:

**Comment by 50bhan** [comment]
@Jibbarth Yeah you are right :laughing: Done :+1:

**Comment by olivernybroe** [comment]
@50bhan Have one bug we need to fix, but then I think this is ready to get merged in.

Running the following command, results in one file scanned
```bash
bin/phpinsights analyse src/Domain/Analyser.php
```

However running it like this, results in 4 files scanned and not 2.
```bash
bin/phpinsights analyse src/Domain/Analyser.php src/Application/Composer.php
```

It looks like it's a bug in `LocalFilesRepository::within`.

**Comment by 50bhan** [comment]
@olivernybroe That's weird! I'll check and let you know.

**Comment by 50bhan** [comment]
@olivernybroe I can't reproduce that! Here is my result:

Screenshot_2020-04-02_17-49-30

**Comment by 50bhan** [comment]
@olivernybroe It was a nice bug actually! You gave these two files:

`bin/phpinsights analyse src/Domain/Analyser.php src/Application/Composer.php`

but the result is:
`src/Application/Composer.php`
`src/Domain/Metrics/Architecture/Composer.php`

`src/Domain/Analyser.php`
`src/Application/Console/Analyser.php`

Because Symfony finder will search in `src/Domain/` and `src/Application/` directories and match with every files with names `Analyser.php` or `Composer.php`. So, I had to filter the final result with full path of files to fix it.

**Comment by olivernybroe** [comment]
@50bhan Just tested it out again, and everything seems to work as expected now!

Nice debugging of the bug, that really was a nasty one of them 👍 

I'll merge this in now. 
Thanks a lot for the awesome contributions you keep making @50bhan ❤️ 

And thanks @Jibbarth for the amazing and always on point reviews!

**Comment by 50bhan** [comment]
@olivernybroe Anytime:) Really happy to contribute! After all I use insights in my personal and work projects, So it's awesome to be a small part of it. Thank you for your usual help and @Jibbarth for amazing reviews.
